Add yagp_hooks_collector to gpcontrib

NJrslv opened the pull request:
Adds [yagp_hooks_collector](https://github.com/open-gpdb/yagp_hooks_collector) to gpcontrib.

I changed make file (gpcontrib/yagp_hooks_collector/Makefile), added a separate test gpcontrib-yagp-hooks-collector in CI that runs collector's tests.

Smyatkin-Maxim committed:
Initial commit: completely useless yet

Just create the overal project structure with a Makefile to
generate protobufs, compile it into a shared library extension
and install it

Smyatkin-Maxim committed:
Add .gitignore

Smyatkin-Maxim committed:
Simple Executor{Start,Finish} logging

as a fundation to build on

Smyatkin-Maxim committed:
Add GRPC into Makefile and some prototype senders

Smyatkin-Maxim committed:
Make ExecuteStart and ExecuteFinish work (partially)

Smyatkin-Maxim committed:
Implement QueryInfo generation

- borrow GlowByte code to generate plan text and SessionInfo
- borrow code from our in-house pg_stat_statements to generate
  query id and plan id
- refactor code to follow common name conventions and identations

Smyatkin-Maxim committed:
Better protobuf filling code

Smyatkin-Maxim committed:
Fix segfault in plan text generator

Smyatkin-Maxim committed:
Add support of spill info and

- do some minor refactoring to follow common naming convention
- add additional message right after ExecutorStart hook

Smyatkin-Maxim committed:
Sync with proto changes + generate normalized query and plan texts

Smyatkin-Maxim committed:
Implement the rest of statistics

1) Query instrumentation
2) /proc/self/* stats

Smyatkin-Maxim committed:
Use llvm code style

Smyatkin-Maxim committed:
Trace events using query_info_collect_hook

It allows finer granularity than executor hooks. Also removed some code
duplication and data duplicaton

Smyatkin-Maxim committed:
Fix two segfaults

1. Initialize query instrumentation to NULL so that it can be properly
checked later (temporary solution, need to find a proper fix)
2. Don't collect spillinfo on query end. Reason: a) it will always be
zero and b) it could crash if we failed to enlarge a spillfile. Seems
like we need some cummulative statistics for spillinfo. Need to check
what explain analyze use.

Smyatkin-Maxim committed:
Don't call hooks from UTILITY mode and increase grpc timeout

Smyatkin-Maxim committed:
More improvements to metrics collector

1. Sync with protobuf changes to collect segment info
2. Remove noisy logging
3. Fix some missing node types in pg_stat_statements

Smyatkin-Maxim committed:
Add debian package configuration

Smyatkin-Maxim committed:
Try older grpcpp headers for bionic builds

Smyatkin-Maxim committed:
Add missing greenplum nodes to pg_stat_statements

Smyatkin-Maxim committed:
Move query stats reporting to ExecutorEnd hook

Reason: when query info hook is called with status 'DONE' planstate
is already deallocated by ExecutorEnd

Smyatkin-Maxim committed:
Change GRPC failure handling

1) Give higher gRPC timeouts to query dispatcher as losing messages
there is more critical
2) If we've failed to send a message via gRPC we notify a background
thread about it and refuse sending any new message until this thread
re-establishes the lost connection

Smyatkin-Maxim committed:
Track CDB metrics and track query nesting

Smyatkin-Maxim committed:
Change package name

Smyatkin-Maxim committed:
Skip nested and utility statements

Smyatkin-Maxim committed:
Slightly rework query skipping

Don't collect system queries with empty query text and ccnt == 0

Smyatkin-Maxim committed:
Get resource groups back into query info

Smyatkin-Maxim committed:
Add some configuration parameters

Smyatkin-Maxim committed:
Capture query execution system stats instead of whole process lifetime

Smyatkin-Maxim committed:
Don't rethrow errors from metrics collector

Rethrowing them might break other extensions and even query execution
pipeline itself

Smyatkin-Maxim committed:
Add forgotten hooks deinitialization

Smyatkin-Maxim committed:
Misc: use modern style error messages

Smyatkin-Maxim committed:
Fix EventSender and GrpcConnector in forks

The problem is both static variable singleton and Grpc don't work well
with forked processes. Delayed their initialization to the actual query
handling time. Should fix the problem.

Smyatkin-Maxim committed:
Set ya-grpc as a dependency

Smyatkin-Maxim committed:
Fix memory leak in analyze code

Smyatkin-Maxim committed:
Add motion and workfile stats

Also fix some compilation issues and minor bugs

Smyatkin-Maxim committed:
Mute all PG-related signals in reconnection thread

Reason: background thread could receive signals, which are supposed to
be handled by the main thread. And only main pg thread knows how to
handle those properly.

Smyatkin-Maxim committed:
Move debian configuration to teamcity

Smyatkin-Maxim committed:
Move logs from psql to log file

also make it more informative with less noice

Smyatkin-Maxim committed:
Some hardening around memory

Smyatkin-Maxim committed:
Remove thread unsafe logging

Smyatkin-Maxim committed:
Dirty hack: make queryid eq planid

Smyatkin-Maxim committed:
Add GUC to filter specific users

Smyatkin-Maxim committed:
Remove dead code and catch thread exception

Smyatkin-Maxim committed:
Several fixes to user filtering

* handle nullptr username properly
* don't start grpc if username is filterred out
* initialize grpc connector with nullptr by default

Smyatkin-Maxim committed:
Minor naming improvement

Smyatkin-Maxim committed:
Add server-side message queueing

Smyatkin-Maxim committed:
Moving to protobuf, part 1: remove GRPC

Smyatkin-Maxim committed:
Add clang-format

Smyatkin-Maxim committed:
Moving to protobuf, part2: initial implementation

Commit:
Pull request #1: YAGP-0.0.1 WIP

Merge in MDB/yagp-hooks-collector from YAGP-0.0.1-WIP to master

* commit '9db45ada73a48ff8095110c205badeaa3419b937': (34 commits)
  Add server-side message queueing
  Minor naming improvement
  Several fixes to user filtering
  Add GUC to filter specific users
  Remove dead code and catch thread exception
  Dirty hack: make queryid eq planid
  Remove thread unsafe logging
  Some hardening around memory
  Move logs from psql to log file
  Move debian configuration to teamcity
  Mute all PG-related signals in reconnection thread
  Add motion and workfile stats
  Fix memory leak in analyze code
  Set ya-grpc as a dependency
  Fix EventSender and GrpcConnector in forks
  Misc: use modern style error messages
  Add forgotten hooks deinitialization
  Don't rethrow errors from metrics collector
  Capture query execution system stats instead of whole process lifetime
  Add some configuration parameters
  ...

Smyatkin-Maxim committed:
Use deprecated protobuf API for bionic compatibility

Smyatkin-Maxim committed:
Replace message queue with incremental message

Should save us some memory footprint, allocations and sends.

Smyatkin-Maxim committed:
Fix bug with missing query statuses

Smyatkin-Maxim committed:
Add runtime statistics on UDS messages

This commit adds a view to see cummulative statistics on UDS messages
sent: total messages, different kinds of errors, max message size.
As a necessary collateral it also renames the extension:
yagp-hooks-collector -> yagp_hooks_collector.
And fixes a bug in UDS socket path resolution.

Smyatkin-Maxim committed:
Move query msg cleanup to the right place

Smyatkin-Maxim committed:
Finalize query msg in DONE hook

Smyatkin-Maxim committed:
Fix some memory leaks

Smyatkin-Maxim committed:
Enable honest query_id collection

Smyatkin-Maxim committed:
Get resource group from current session

Instead of getting default resgroup for current role. Reason: resource
group can be reassigned in another extension via resgroup_assign_hook.

Smyatkin-Maxim committed:
Added support for nested queries

- this commit allows to track or not track nested queries
- presumably fixes some bad bugs when we get wrong message sequence. For
  example, sending q1`s END state for q2 instead. There have been
multiple bugs like this

Smyatkin-Maxim committed:
Trim long text fields

Smyatkin-Maxim committed:
Report error messages for failed queries

Smyatkin-Maxim committed:
Don't track the CANCELLING event

Smyatkin-Maxim committed:
Properly send runtime metrics on CANCEL/ERROR

Smyatkin-Maxim committed:
Allow anyone to disable report_nested_queries

Smyatkin-Maxim committed:
Diff stats between executor start and end

The reason is that for nested statements any incremental stats (spills,
proc stats, execution time) have to be calculated separately.

Smyatkin-Maxim committed:
Fix try/catch block when calling cpp code

Smyatkin-Maxim committed:
Don't normalize trimmed plans

Smyatkin-Maxim committed:
Clean up forgotten text fields

Smyatkin-Maxim committed:
[MDB-31938] Send nested queries only from master

Smyatkin-Maxim committed:
[MDB-31936] Add slice info

Smyatkin-Maxim committed:
Send nested queries summary stats from segments

Smyatkin-Maxim committed:
[Refactoring] Split EventSender into submodules

Smyatkin-Maxim committed:
Ignore EXPLAIN VERBOSE errors

leborchuk committed:
Merge pull request #1 from open-gpdb/max/ignore-explain-errors

Ignore EXPLAIN VERBOSE errors

Smyatkin-Maxim committed:
Add support for per-slice interconnect statistics

This change requires a patched version of gp code which provides
interconnect teardown hook

Smyatkin-Maxim committed:
Don't rely on IC hook for compilation

reshke committed:
Merge pull request #2 from open-gpdb/max/query_ic_stats

 Add per-slice UDP interconnect statistics

NJrslv committed:
fix: propagate ignored users on update

NJrslv committed:
propagate ignored users only when executor starts

NJrslv committed:
fix ub in strcpy

NJrslv committed:
refactor

NJrslv committed:
Merge pull request #5 from NJrslv/fix-filter-ignored-users-on-upd

fix: propagate ignored users on update

NJrslv committed:
Merge pull request #6 from NJrslv/fix-ub-strcpy

fix ub in strcpy

NJrslv committed:
Add EXPLAIN ANALYZE metrics collection

This commit depends on a new hook in gpdb (but can be compiled w/o it) and allows sending
analyze output for long-running queries.

NJrslv committed:
parallel makefile & move link flags

NJrslv committed:
parallel makefile & move link flags (#8)

NJrslv committed:
correct expand var

NJrslv committed:
Merge branch 'YAGP-0.0.2-WIP' into fix-makefile

NJrslv committed:
Merge pull request #9 from open-gpdb/fix-makefile

Fix makefile

NJrslv committed:
add safe cpp wrappers around pg funcs

NJrslv committed:
add minimal readme

NJrslv committed:
add minimal readme

NJrslv committed:
add nested queries info

NJrslv committed:
add nested queries info

NJrslv committed:
move nested queries to general info

NJrslv committed:
move nested queries to general info

NJrslv committed:
log actual err msg & templatize code & make free noexcept

NJrslv committed:
add tab

NJrslv committed:
change namespace name to avoid conflicts with gpos

NJrslv committed:
move template to .cpp

NJrslv committed:
fix mem leak

NJrslv committed:
Merge pull request #14 from open-gpdb/fix-memleak

fix mem leak

NJrslv committed:
Merge pull request #11 from open-gpdb/wrap-pg-calls

add safe cpp wrappers around pg funcs

NJrslv committed:
add trimming gucs to readme

NJrslv committed:
Merge branch 'YAGP-0.0.2-WIP' into minimal-readme

NJrslv committed:
fix typo

NJrslv committed:
Merge pull request #15 from open-gpdb/minimal-readme

add trimming gucs to readme

NJrslv committed:
add metrics desc

NJrslv committed:
clear

NJrslv committed:
Merge pull request #17 from open-gpdb/metric-doc

Metrics doc

NJrslv committed:
Fix missing states (utility statements included) (#16)

NJrslv committed:
fix cpp & pg memleak

NJrslv committed:
Merge pull request #18 from open-gpdb/fix-memleak

fix cpp & pg memleak

NJrslv committed:
report bug queries at the end of extension (#19)

* ereport(log) bug queries at the end of extension

NJrslv committed:
fix defs & trim spaces (#21)

NJrslv committed:
Add PG alike tests (#20)

* add pg alike tests

NJrslv committed:
send analyze in text & enable it (#22)

* send analyze in text & enable it

NJrslv committed:
clean (#23)

rm -f gmon.out & fix typo

NJrslv committed:
report utility stmt (#24)

* report utility stmt

NJrslv committed:
Trim utf8 (#25)

Trim strings larger than 1MB by default and if we cut multi-byte utf8 then discard the character and shift the cut position.

andr-sokolov commented on .github/workflows/build-gpdb.yml:
```diff
@@ -795,6 +801,7 @@ jobs:
 
           apt update -y
           apt install -y tzdata
+          apt install -y libprotobuf-dev protobuf-compiler
```
What about combining this line with the previous one to don't run `apt install` twice?

NJrslv replied:
I wanted to highlight different responsibilities: the first line is only about tzdata, and the second is about protobuf and only. Do you think I still need to merge them? :)

andr-sokolov replied:
What about one command, but two lines?
```
apt install -y tzdata \
               libprotobuf-dev protobuf-compiler
```

NJrslv replied:
got it, thanks

andr-sokolov commented on gpcontrib/yagp_hooks_collector/Makefile:
```diff
@@ -0,0 +1,47 @@
+override CFLAGS = -Wall -Wmissing-prototypes -Wpointer-arith -Wendif-labels \
+	-Wmissing-format-attribute -Wformat-security -fno-strict-aliasing -fwrapv \
+	-fexcess-precision=standard -Wno-unused-but-set-variable -Wno-address \
+	-Wno-format-truncation -Wno-stringop-truncation -g -ggdb -std=gnu99 \
```
Do we really always need debug symbols?

NJrslv replied:
I don't know if we need debug symbols in the release version.

On the one hand, since the hooks collector was drastically refactored, the symbols will help with bugs and core dumps in releases.

On the other hand, fewer compiler optimizations are performed, but I think the cost of it is negligible compared to the cost of networking accesses in Greenplum and all those things.

And most importantly there are 2-3 .c files.

What do you think?

andr-sokolov replied:
usually CFLAGS are taken from pg_config to use the same flags as in the core. CXXFLAGS contains `g -ggdb` too

NJrslv replied:
okay, what do you think if I just remove the `CFLAGS` from makefile and leave `CXXFLAGS` unchanged? Does it make sense?

NJrslv commented:
Please, don't merge. It's better to copy the extension as submodule and disable it by default.
Or we can leave it as it is but instead update commit's history. I am doing the latter.
Upd: now it is good to go

NJrslv committed:
Add 'gpcontrib/yagp_hooks_collector/' from commit 'ca620e9f75f5bc717620d529ef6753c7666c2d99'

git-subtree-dir: gpcontrib/yagp_hooks_collector
git-subtree-mainline: 00c7301aeeb60743f46de2fdedefdede135bcd16
git-subtree-split: ca620e9f75f5bc717620d529ef6753c7666c2d99

NJrslv commented:
Do we want to pollute `OPENGPDB_STABLE` branch history?

NJrslv committed:
add test to CI & --with-yagp-hooks-collector opt

correct & add ci test

add hooks collector to gpcontrib makefile

add configure option, without by default

add missing flag to ci to run the test

NJrslv commented on gpcontrib/yagp_hooks_collector/expected/yagp_cursors.out:
```diff
@@ -55,7 +56,7 @@ RESET yagpcc.logging_mode;
 SELECT segid, query_text, query_status FROM yagpcc.log WHERE segid = -1 AND utility = true ORDER BY segid, ccnt, yagp_status_order(query_status) ASC;
  segid |                      query_text                       |    query_status     
 -------+-------------------------------------------------------+---------------------
-    -1 |                                                       | QUERY_STATUS_DONE
```
As you can see, the test output changed - here’s why.

`yagp_hooks_collector` sends data to another process via UDS sockets through `UDSConnector::report_query()`. If no target process exists, the function returns `false`.

When the data is sent successfully (`true` is returned from `report_query()`), large fields like the query text are cleared, otherwise, they remain. In the test, the query text now appears because, in the CI container, there’s no receiving process - so the field isn’t reset.

This behavior is intentional. We collect query data in four stages, and if one fails, later stages can still transmit large fields.

NJrslv committed:
Merge branch 'OPENGPDB_STABLE' into yagp_hooks_collector
